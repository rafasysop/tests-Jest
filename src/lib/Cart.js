import { remove } from "lodash";
import find from "lodash/find";

export class Cart {
  items = [];
  add(item) {
    const itemToFind = { product: item.product };
    const hasProduct = find(this.items, itemToFind);

    if (hasProduct) {
      remove(this.items, itemToFind);
    }

    this.items.push(item);
  }
  getTotal() {
    return this.items.reduce(
      (acc, item) => acc + item.product.price * item.quantity,
      0
    );
  }
  remove(product) {
    remove(this.items, { product });
  }
}
