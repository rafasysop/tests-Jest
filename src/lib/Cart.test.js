import { Cart } from "./Cart";

describe("Cart", () => {
  let cart;
  let product = {
    title: "Echo Dot",
    price: 35388, //353.88
  };

  let product2 = {
    title: "Echo Show",
    price: 41872, //418.72
  };

  beforeEach(() => {
    cart = new Cart();
  });

  it("should return 0 when getTotal() is executed in a newly created instance", () => {
    const cart = new Cart();
    expect(cart.getTotal()).toEqual(0);
  });

  it("should multiply quantity and price the total amount", () => {
    cart.add({
      product,
      quantity: 2, //70776
    });
    expect(cart.getTotal()).toEqual(70776);
  });

  it("should ensure no more than on product exists at a time", () => {
    cart.add({
      product,
      quantity: 2, //70776
    });
    cart.add({
      product,
      quantity: 1, //70776
    });
    expect(cart.getTotal()).toEqual(35388);
  });

  it("should update total when a product gets included and then removed", () => {
    cart.add({
      product,
      quantity: 2, //70776
    });
    cart.add({
      product: product2,
      quantity: 1, //41872
    });

    cart.remove(product);
    expect(cart.getTotal()).toEqual(41872);
  });
});
