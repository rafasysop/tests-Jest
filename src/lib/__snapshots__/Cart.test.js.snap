// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Checkout Cart should  1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Echo Dot",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Echo Show",
      },
      "quantity": 1,
    },
  ],
  "total": 112648,
}
`;
